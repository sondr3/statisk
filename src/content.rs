use std::path::{Path, PathBuf};

use anyhow::{Context, Result, anyhow, bail};
use minijinja::{context, value::Value};
use serde::Serialize;

use crate::{
    context::StatiskContext as SContext,
    frontmatter::Frontmatter,
    jotdown::render_jotdown,
    templating::{TemplatePath, create_base_context},
    typst::render_typst,
    utils::{split_frontmatter, unprefixed_parent},
};

#[allow(clippy::upper_case_acronyms)]
#[derive(Debug, Copy, Clone, Serialize)]
pub enum ContentType {
    HTML,
    XML,
    Jotdown,
    Typst,
    Unknown,
}

impl ContentType {
    pub fn from_ext(path: &Path) -> Result<Self> {
        match path.extension() {
            None => bail!("No extension for content type"),
            Some(kind) => match kind.to_string_lossy().to_string().as_ref() {
                "xml" | "xsl" => Ok(ContentType::XML),
                "html" => Ok(ContentType::HTML),
                "typ" => Ok(ContentType::Typst),
                "dj" => Ok(ContentType::Jotdown),
                _ => Ok(ContentType::Unknown),
            },
        }
    }
}

#[derive(Debug, Serialize, Clone)]
pub struct Content {
    pub source: PathBuf,
    pub out_path: PathBuf,
    pub dir: Option<String>,
    pub url: String,
    pub frontmatter: Frontmatter,
    pub content: String,
    pub kind: ContentType,
}

impl Content {
    pub fn from_path(path: &Path, root: &Path, kind: ContentType) -> Result<Self> {
        let file = std::fs::read_to_string(path)?;
        let stem = path.file_stem().unwrap().to_string_lossy();
        let stem = stem.as_ref();

        let (frontmatter, content) =
            split_frontmatter(&file).ok_or(anyhow!("Could not find content or frontmatter"))?;

        let frontmatter = match (kind, frontmatter) {
            (ContentType::XML, None) => Frontmatter::empty(),
            (_, Some(fm)) => Frontmatter::deserialize(&fm)?,
            _ => bail!("Missing frontmatter in content"),
        };

        let dir = unprefixed_parent(path, root);
        let out_path = out_path(kind, path, dir.as_ref(), stem, &frontmatter);

        let url = frontmatter.url(
            &out_path
                .parent()
                .map(|p| p.to_string_lossy().to_string())
                .unwrap_or_default(),
        );

        Ok(Content {
            source: path.to_path_buf(),
            url,
            kind,
            out_path,
            dir,
            content,
            frontmatter,
        })
    }

    pub fn render(&self, context: &SContext) -> Result<String> {
        match self.kind {
            ContentType::HTML | ContentType::XML => self.render_template(context),
            ContentType::Jotdown | ContentType::Typst => self.render_content(context),
            ContentType::Unknown => bail!("Cannot render unknown files"),
        }
    }

    pub fn filename(&self) -> String {
        self.source.file_name().map_or_else(
            || panic!("No filename found"),
            |name| name.to_string_lossy().to_string(),
        )
    }

    pub fn is_public_page(&self) -> bool {
        matches!(self.kind, ContentType::Jotdown | ContentType::HTML) && !self.is_special_page()
    }

    pub fn context(&self, context: &SContext) -> Result<Value> {
        let content = match self.kind {
            ContentType::Jotdown => render_jotdown(&self.content)?,
            ContentType::Typst => render_typst(&self.content)?,
            _ => self.content.clone(),
        };
        let frontmatter_context = self.frontmatter.to_context();

        Ok(context! {
            ..frontmatter_context,
            ..context! {
                content => content,
                canonical_url => context.statisk.url().join(&self.url)?,
            }
        })
    }

    fn is_special_page(&self) -> bool {
        self.out_path == PathBuf::from("404.html") || self.out_path == PathBuf::from("500.html")
    }

    fn layout(&self) -> TemplatePath {
        match &self.frontmatter.layout {
            Some(layout) => TemplatePath(None, layout.to_string()),
            None => TemplatePath(self.dir.clone(), "page".to_string()),
        }
    }

    fn render_content(&self, app_context: &SContext) -> Result<String> {
        let base_context = create_base_context(app_context);
        let context = self.context(app_context)?;
        let context = context! { ..base_context, ..context };
        app_context
            .templates
            .render_template(&self.layout(), context)
    }

    fn render_template(&self, app_context: &SContext) -> Result<String> {
        let base_context = create_base_context(app_context);
        let context = self.context(app_context)?;
        let context = context! { ..base_context, ..context };
        let env = app_context.templates.environment.acquire_env()?;
        let template = env.template_from_str(&self.content)?;
        template.render(context).context("Could not render")
    }
}

fn out_path(
    kind: ContentType,
    path: &Path,
    dir: Option<&String>,
    stem: &str,
    frontmatter: &Frontmatter,
) -> PathBuf {
    match kind {
        ContentType::XML => PathBuf::from(path.file_name().unwrap_or_default()),
        ContentType::HTML | ContentType::Unknown => {
            // First check if this is a special page (like 404.html)
            if let Some(filename) = path.file_name().and_then(|f| f.to_str()) {
                if filename == "404.html" || filename == "500.html" {
                    return PathBuf::from(filename);
                }
            }

            // Then handle regular pages
            match (dir, &frontmatter.slug) {
                (None, None) => PathBuf::from("index.html"),
                (None, Some(slug)) => [slug, "index.html"].into_iter().collect(),
                (Some(dir), Some(slug)) => [dir, slug, "index.html"].into_iter().collect(),
                (Some(dir), None) => [dir, "index.html"].into_iter().collect(),
            }
        }
        ContentType::Jotdown | ContentType::Typst => match &frontmatter.slug {
            Some(slug) => [slug, "index.html"].into_iter().collect(),
            None => [stem, "index.html"].into_iter().collect(),
        },
    }
}
