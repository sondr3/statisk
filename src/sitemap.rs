use std::{
    fmt::{self, Display, Formatter},
    io::Write,
};

use anyhow::{Context, Result};
use jiff::civil::Date;
use url::Url;
use xml::{common::XmlVersion, writer::XmlEvent, EmitterConfig, EventWriter};

use crate::content::Content;

#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq)]
pub enum ChangeFreq {
    Always,
    Hourly,
    Daily,
    Weekly,
    Monthly,
    Yearly,
    Never,
}

impl Display for ChangeFreq {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            ChangeFreq::Always => write!(f, "always"),
            ChangeFreq::Hourly => write!(f, "hourly"),
            ChangeFreq::Daily => write!(f, "daily"),
            ChangeFreq::Weekly => write!(f, "weekly"),
            ChangeFreq::Monthly => write!(f, "monthly"),
            ChangeFreq::Yearly => write!(f, "yearly"),
            ChangeFreq::Never => write!(f, "never"),
        }
    }
}

#[derive(Debug)]
pub struct UrlEntry {
    pub loc: Url,
    pub last_mod: Option<Date>,
    pub change_freq: Option<String>,
    pub priority: Option<f32>,
}

impl UrlEntry {
    pub fn new(
        loc: Url,
        last_mod: Option<Date>,
        change_freq: Option<ChangeFreq>,
        priority: Option<f32>,
    ) -> Self {
        assert!(priority.map_or(true, |e| (0.0..=1.0).contains(&e)));

        Self {
            loc,
            last_mod,
            change_freq: change_freq.map(|e| e.to_string()),
            priority,
        }
    }

    pub fn from_content(value: &Content, base: &Url) -> Result<Self> {
        let url = base.join(&value.url)?;

        Ok(UrlEntry::new(
            url,
            value.frontmatter.last_modified,
            Some(ChangeFreq::Monthly),
            None,
        ))
    }

    pub fn to_xml<W: Write>(&self, writer: &mut EventWriter<W>) -> Result<()> {
        writer.write(XmlEvent::start_element("url"))?;

        writer.write(XmlEvent::start_element("loc"))?;
        writer.write(&*self.loc.to_string())?;
        writer.write(XmlEvent::end_element())?;

        if let Some(last_mod) = &self.last_mod {
            writer.write(XmlEvent::start_element("lastmod"))?;
            writer.write(&*last_mod.to_string())?;
            writer.write(XmlEvent::end_element())?;
        }
        if let Some(change_freq) = &self.change_freq {
            writer.write(XmlEvent::start_element("changefreq"))?;
            writer.write(change_freq.as_str())?;
            writer.write(XmlEvent::end_element())?;
        }
        if let Some(priority) = &self.priority {
            writer.write(XmlEvent::start_element("priority"))?;
            writer.write(priority.to_string().as_str())?;
            writer.write(XmlEvent::end_element())?;
        }

        writer.write(XmlEvent::end_element())?;
        Ok(())
    }
}

pub fn create(urls: Vec<UrlEntry>) -> Result<String> {
    let mut res = Vec::new();
    let mut writer = EmitterConfig::new()
        .perform_indent(true)
        .create_writer(&mut res);

    let doc = XmlEvent::StartDocument {
        version: XmlVersion::Version10,
        encoding: Some("UTF-8"),
        standalone: None,
    };

    writer.write(doc)?;

    let stylesheet = XmlEvent::processing_instruction(
        "xml-stylesheet",
        Some(r#"type="text/xsl" href="/sitemap.xsl"?"#),
    );

    writer.write(stylesheet)?;

    let url_set = XmlEvent::start_element("urlset")
        .ns("", "http://www.sitemaps.org/schemas/sitemap/0.9")
        .ns("image", "http://www.google.com/schemas/sitemap-image/1.1")
        .ns("video", "http://www.google.com/schemas/sitemap-video/1.1");

    writer.write(url_set)?;

    for url in urls {
        url.to_xml(&mut writer)?;
    }

    writer.write(XmlEvent::end_element())?;

    String::from_utf8(res).context("Sitemap failed to serialize")
}

#[cfg(test)]
mod tests {
    use jiff::civil::Date;
    use url::Url;

    use crate::sitemap::{create, ChangeFreq, UrlEntry};

    #[test]
    fn test_sitemap() {
        let urls = vec![
            UrlEntry::new(
                Url::parse("http://www.example.com/").unwrap(),
                None,
                None,
                None,
            ),
            UrlEntry::new(
                Url::parse("https://example.org/").unwrap(),
                Some(Date::new(2005, 1, 1).unwrap()),
                Some(ChangeFreq::Monthly),
                Some(0.8),
            ),
            UrlEntry::new(
                Url::parse("http://www.example.com/catalog?item=12&amp;desc=vacation_hawaii")
                    .unwrap(),
                None,
                Some(ChangeFreq::Weekly),
                None,
            ),
            UrlEntry::new(
                Url::parse("http://www.example.com/catalog?item=73&amp;desc=vacation_new_zealand")
                    .unwrap(),
                Some(Date::new(2004, 12, 23).unwrap()),
                Some(ChangeFreq::Weekly),
                None,
            ),
            UrlEntry::new(
                Url::parse("http://www.example.com/catalog?item=74&amp;desc=vacation_newfoundland")
                    .unwrap(),
                Some(Date::new(2004, 12, 23).unwrap()),
                None,
                Some(0.3),
            ),
            UrlEntry::new(
                Url::parse("http://www.example.com/catalog?item=83&amp;desc=vacation_usa").unwrap(),
                Some(Date::new(2004, 11, 23).unwrap()),
                None,
                None,
            ),
        ];

        let sitemap = create(urls).unwrap();
        insta::assert_snapshot!(sitemap);
    }
}
