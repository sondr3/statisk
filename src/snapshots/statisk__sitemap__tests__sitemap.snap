---
source: src/sitemap.rs
expression: sitemap
---
<?xml version="1.0" encoding="UTF-8"?>
<?xml-stylesheet type="text/xsl" href="/sitemap.xsl"??>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:image="http://www.google.com/schemas/sitemap-image/1.1" xmlns:video="http://www.google.com/schemas/sitemap-video/1.1">
  <url>
    <loc>http://www.example.com/</loc>
  </url>
  <url>
    <loc>https://example.org/</loc>
    <lastmod>2005-01-01</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.8</priority>
  </url>
  <url>
    <loc>http://www.example.com/catalog?item=12&amp;amp;desc=vacation_hawaii</loc>
    <changefreq>weekly</changefreq>
  </url>
  <url>
    <loc>http://www.example.com/catalog?item=73&amp;amp;desc=vacation_new_zealand</loc>
    <lastmod>2004-12-23</lastmod>
    <changefreq>weekly</changefreq>
  </url>
  <url>
    <loc>http://www.example.com/catalog?item=74&amp;amp;desc=vacation_newfoundland</loc>
    <lastmod>2004-12-23</lastmod>
    <priority>0.3</priority>
  </url>
  <url>
    <loc>http://www.example.com/catalog?item=83&amp;amp;desc=vacation_usa</loc>
    <lastmod>2004-11-23</lastmod>
  </url>
</urlset>
